src/Writers/HTMLWriter.jl

```diff
@@ -813,30 +813,27 @@ function pretty_url(ctx, path::AbstractString)
 end
 
 """
-Tries to guess the page title by looking at the `<h1>` headers and returns the
-header contents of the first `<h1>` on a page (or `nothing` if the algorithm
-was unable to find any `<h1>` headers).
+Tries to guess the page title by first looking at the `<h1>` headers and next at `<h2>`
+headers. Returns the contents of the first `<h1>`/`<h2>` on the page.
+If no `<h1>` or `<h1>` headings ar found we fall back to the filename of the page.
 """
 function pagetitle(page::Documents.Page)
-    title = nothing
-    for element in page.elements
-        if isa(element, Markdown.Header{1})
-            title = element.text
-            break
-        end
-    end
-    title
+    elements = page.elements
+    idx = Compat.findfirst(x -> isa(x, Markdown.Header{1}), elements)
+    idx !== nothing && return elements[idx].text
+    idx = Compat.findfirst(x -> isa(x, Markdown.Header{2}), elements)
+    idx !== nothing && return elements[idx].text
+    return basename(page.source)
 end
 
 function pagetitle(ctx, navnode::Documents.NavNode)
     navnode.title_override === nothing || return navnode.title_override
 
     if navnode.page !== nothing
-        title = pagetitle(getpage(ctx, navnode))
-        title === nothing || return title
+        return pagetitle(getpage(ctx, navnode))
     end
 
-    "-"
+    "-" # unreachable?
```

Comment:
In practice it should be since `navnode.page === nothing` can only happen when there's a `title_override`. Could perhaps invert the `if` and have a nice `error()` if `navnode.page === nothing` (in case someone runs into case when hacking on the internals etc).